충남대 BE_좌진우 3단계 - 상품 옵션

jinwoojwa opened the pull request:
## Step 3 - 상품 옵션

✅ 주요 변경 사항

- 상품에 옵션 추가 기능 구현
- 옵션은 이름, 수량 정보를 포함하며 상품에 1개 이상 존재해야 함
- 옵션 이름 유효성 검증 로직 구현
- 상품 내 옵션 이름 중복 불가 처리
- 옵션 수량 차감 기능 추가
- OptionNameTest, OptionQuantityTest 클래스 구현을 통한 단위 테스트 추가

<br>

🧩 테스트

- `OptionNameTest`: 옵션 이름의 유효성 검증 테스트
- `OptionQuantityTest`: 옵션 수량의 범위 및 차감 기능 테스트

<br>

❓

미션 기능 요구 사항에 옵션에 대한 CRUD 기능을 모두 구현해야 한다라는 요구사항은 없는 것 같아

우선 옵션 추가와, 옵션 조회에 대한 API만을 구현했습니다. 추가로 구현해야 할 사항이 있으면 말씀해주시면

감사하겠습니다!!

jinwoojwa committed:
docs(README): 상품 옵션 기능 요구 사항 추가

jinwoojwa committed:
feat: 상품 옵션(Option) 도메인 엔티티 및 객체 도입

- 상품 옵션 엔티티 Option 생성
- 옵션명(OptionName), 옵션 수량(OptionQuantity) 값 객체로 분리하여
  도메인 구현
- 옵션 수량 감소 기능 및 유효성 검증 로직 포함
- Product 엔티티와 다대일(ManyToOne) 관계 설정

jinwoojwa committed:
feat: Product 엔티티에 Option 연관관계 매핑 추가

jinwoojwa committed:
feat: 데이터 초기화 로직에 옵션(Option) 초기 데이터 추가

jinwoojwa committed:
feat: 상품 옵션 목록 조회 API 구현

- 특정 상품에 대한 옵션 목록을 조회하는 GET /api/products/{productId}/options API 추가
- OptionController에 GET 메서드 추가
- OptionService에 getOptionsByProductId 메서드 구현
- OptionRepository에 findByProduct 메서드 정의
- OptionResponseDto 생성 및 매핑 로직 추가

jinwoojwa committed:
feat: 상품에 옵션 추가 API 구현

jinwoojwa committed:
test: OptionName, OptionQuantity 값 객체 단위 테스트 추가

jinwoojwa committed:
docs(README): 세부 구현 내용 추가

jinwoojwa committed:
fix: Option 조회 시 N+1 문제 해결을 위한 @EntityGraph 적용

liquidjoo commented on README.md:
```diff
@@ -31,4 +32,43 @@
+- API 구현
+  + `GET /api/products/{productId}/options`
+    * 특정 상품의 모든 옵션 목록을 조회하는 API 추가
+  + `POST /api/products/{productId}/options`
+    * 상품에 옵션을 추가하는 API 추가
+    * 입력값 검증 및 중복 이름 검사 후 저장
```
상품에는 항상 하나 이상의 옵션이 존재해야 함
이라는 제약 조건을 보았을 때, options 의 생성 방법은 어떻게 생성하는 것이 가장 자연스러울까요?

jinwoojwa replied:
상품 생성 시점에 옵션과 함께 생성되어야 합니다. 상품에 대한 서비스 코드에서 옵션이 존재하지 않을 때, 예외를 발생시키도록

처리하여 "상품은 반드시 하나 이상의 옵션을 가져야 한다"라는 제약 조건을 강제하도록 변경했습니다.

또한, 옵션을 상품 생성 이후에 따로 추가하는 방식은 상품 객체가 일시적으로 옵션을 가지지 않는 상태로 존재할 수 있었습니다.

따라서, 상품 생성 시점에 옵션 리스트를 전달 받아 내부적으로 `addOption()`을 호출하여 옵션과의 연관관계를 만족하도록 했습니다.

liquidjoo commented on src/main/java/gift/common/config/DataInitializer.java:
```diff
@@ -13,19 +15,37 @@ public class DataInitializer implements CommandLineRunner {
+        optionRepository.save(option1);
+        optionRepository.save(optton2);
+        optionRepository.save(option3);
+        optionRepository.save(option4);
```
상품을 생성 시에 하나의 옵션 이상을 필수로 가져야하는데 어떤 흐름이 가장 자연스러울까요?
상품을 만들고, 상품 정보를 받아서 옵션을 저장하는 것도 좋지만, 연관관계를 잘 활용하여 한번에 생성하는 것도 좋지 않을까요?

jinwoojwa replied:
```java
Product coldBrew = new Product("콜드브루", 4500, "https://image.istarbucks.co.kr/upload/store/skuimg/2025/06/[9200000000038]_20250626095744579.jpg");
        Product americano = new Product("아메리카노", 4000, "https://image.istarbucks.co.kr/upload/store/skuimg/2025/06/[110563]_20250626094354080.jpg");
        Product cappuccino = new Product("카푸치노", 5000, "https://image.istarbucks.co.kr/upload/store/skuimg/2021/04/[38]_20210415154821991.jpg");

        americano.addOption(Option.of("Tall", 10, americano));
        americano.addOption(Option.of("Grande", 5, americano));
        coldBrew.addOption(Option.of("Tall", 3, coldBrew));
        cappuccino.addOption(Option.of("Venti", 2, cappuccino));

        productRepository.save(coldBrew);
        productRepository.save(americano);
        productRepository.save(cappuccino);
```

위와 같이 `save()`를 한번만 수행하도록 변경했습니다.

하지만, 저는 이 방법도 여전히 상품 객체가 옵션을 가지지 않는 상태로 존재하는 순간이 있다고 생각합니다.

위 방법은 영속성 컨텍스트에 진입하기 전 상태이므로 괜찮은 걸까요???

liquidjoo replied:
연관관계가 있다면 차라리 Product 생성 시에 바로 받아볼 수 있지 않을까요?

jinwoojwa replied:
```java
Product coldBrew = Product.createProduct(
                "콜드브루", 4500,
                "https://image.istarbucks.co.kr/upload/store/skuimg/2025/06/[9200000000038]_20250626095744579.jpg",
                List.of(
                        new OptionRequestDto("Tall", 3)
                )
        );

        Product americano = Product.createProduct(
                "아메리카노", 4000,
                "https://image.istarbucks.co.kr/upload/store/skuimg/2025/06/[110563]_20250626094354080.jpg",
                List.of(
                        new OptionRequestDto("Tall", 10),
                        new OptionRequestDto("Grande", 5)
                )
        );
```

Product 생성 시점에 옵션까지 함께 받아 생성하는 것이 바람직한 방식인 것 같습니다. 위 코드와 같이 변경 완료했습니다.

liquidjoo commented on src/main/java/gift/option/entity/Option.java:
```diff
@@ -0,0 +1,56 @@
+        this.product = product;
+    }
+
+    public static Option of(String name, int quantity, Product product) {
+        return new Option(new OptionName(name), new  OptionQuantity(quantity), product);
+    }
```
정적메서드패턴을 사용한 이유가 있을까요?
어떠한 경우에 이러한 패턴을 사용하면 좋을지 고민해보면 좋을 것 같아요!

jinwoojwa replied:
정적 팩토리 메서드를 사용한 이유는 `OptionName`과 `OptionQuantity`의 생성자 호출을 직접 처리하지 않고, Option 객체를 쉽게 생성할 수 있도록 하기 위해 사용했습니다. 이러한 세부 구현을 캡슐화하여 코드의 가독성을 높이는 데도 좋다고 생각했습니다.

liquidjoo replied:
생성자 또한 캡슐화가 가능한 것 같은데요 그렇다면 충분히 생성자를 활용해도 괜찮지 않을까요?
정적 팩토리 메서드가 가지는 이점이 무엇이 있을까요?
> 정적 팩토리 메서드가 좋아보일 수 있지만 왜 이러한 패턴이 나오게 되었는지 고민해 보면 좋습니다

jinwoojwa replied:
현재의 상태에서는 굳이 정적 팩토리 메서드를 사용할 필요가 없는 것 같습니다. 말씀해주신대로 현재의 구현에서는 단순히 여러 번의 생성자 호출에서 new OptionName(name), new OptionQuantity(quantity)의 사용을 줄이는데만 의의가 있다고 생각합니다. 

만약, 이후 요구사항이 확장되면서 옵션 생성 시 유효성 검증이 추가되거나, 특정 조건에 따라 옵션 생성을 제한하는 등의 로직이 필요해지는 경우 정적 팩토리 메서드가 생성자보다 유연하게 변경사항에 대응할 수 있을 것이라 생각합니다.

정적 팩토리 메서드에 대한 학습을 더욱 진행해보겠습니다!

liquidjoo commented on src/main/java/gift/option/service/DefaultOptionService.java:
```diff
@@ -0,0 +1,58 @@
+    @Override
+    public List<OptionResponseDto> getOptionsByProductId(Long productId, Long memberId) {
+
+        Product product = productRepository.findById(productId)
+                .orElseThrow(() -> new ProductNotFoundException(productId));
+
+        List<Option> options = optionRepository.findByProduct(product);
```
상품과 옵션은 어떤 관계를 갖고있을까요?
상품내에 옵션을 같은 컨텍스트로 이해할 것인가, 아니면 다른 컨텍스트로 이해할 것 인가에 대해서도 고민해보면 좋을 것 같네요!

jinwoojwa replied:
처음 미션을 수행할 때부터 했던 고민인 것 같습니다.

저는 옵션이 상품에 종속적이며, 상품의 내부 구성 요소 중 하나라고 생각했습니다.

상품이 생성될 때 옵션도 함께 생성되고, 상품이 삭제되면 옵션도 같이 삭제되어야 하기 때문에 같은 도메인 컨텍스트 내에서

내부 구성 요소로 이해해야 한다고 판단했습니다. 따라서 옵션을 `Product` 엔티티의 필드 구성 요소 중  하나로 구현하는 방식도

고민했었습니다.

하지만, 실제 프로젝트에서는 `Product` 패키지의 크기가 너무 커져 관리가 어려워질 것 같아 옵션 도메인을 별도의 패키지로

분리하여 구현했습니다. 이 과정에서 연관관계를 제대로 설정하지 않아 문제가 발생했던 것 같습니다.

<br>

혹시, 리뷰어님이라면 상품과 옵션 도메인을 같은 컨텍스트 내에서 관리하는 것과 별도 패키지로 분리하는 것 중 어떤 방식을

추천하시는지 궁금합니다

liquidjoo replied:
조금 더 쉽게 접근하면 `옵션만 존재할 수 있는가?` 라고 생각해 본다면 어떨까요?
더 나아가 패키지의 크기가 커지더라도 이해와 유추를 잘 할 수 있도록 분리하면 패키지의 크기는 문제가 되지 않을거라 생각합니다
실무자 입장에선 한 패키지에 수십개 이상의 파일이 존재할 수도 있기 때문입니다!
그래서 단일책임원칙과 이해하기쉬운 코드, 메서드 명, 클래스 명을 중요시 하는 이유도 있습니다

liquidjoo reviewed (changes requested):
안녕하세요 진우님
미션 반영 잘해주셨어요 👍 
몇 가지 코멘트 남겼습니다 확인하고 다시 요청해 주세요 🙇

jinwoojwa committed:
refactor: DataInitializer에서 상품과 옵션의 연관관계를 반영하여 초기 데이터 주입 개선

jinwoojwa committed:
feat: 상품 생성 시 옵션 필수 조건 검증 및 관련 예외 처리 추가

- ProductOptionRequiredException 커스텀 예외 클래스 생성
- 옵션 필수 조건 위반 시 예외 발생하도록 서비스 로직 수정
- 글로벌 예외 처리기에 ProductOptionRequiredException 핸들러 추가
- 옵션 없는 상품 생성 요청 시 400 Bad Request와 명확한 에러 메시지 반환

jinwoojwa commented:
## 💻 테스트

- ❌ 상품 추가 테스트 (옵션 X)
<img width="400" height="350" alt="image" src="https://github.com/user-attachments/assets/be9ca7ca-e271-4aea-a26a-7409eb32b9b6" />

---

- ✅ 상품 추가 테스트 (옵션 O)
<img width="614" height="613" alt="image" src="https://github.com/user-attachments/assets/8d058bfc-09bf-4889-9931-0f6f5a43c589" />

---

- ✅ 추가한 상품 옵션 조회
<img width="485" height="474" alt="image" src="https://github.com/user-attachments/assets/a633cbee-f0aa-402a-b9de-348c13563baa" />

liquidjoo commented on src/main/java/gift/option/entity/OptionName.java:
```diff
@@ -0,0 +1,34 @@
+
+import java.util.regex.Pattern;
+
+@Embeddable
```
Embeddable 활용 좋네요
원시 값을 포장하여 객체를 다루게 된다면 많은 이점이 존재하는데
후에 잘 활용할 수 있게 되면 좋을 것 같아요~

liquidjoo reviewed (changes requested):
안녕하세요 진우님
피드백 반영 잘해주셨어요 👍 
질문 주신 내용에 대한 답변을 달아놓았는데요
머지해도 되지만 진우님의 의견을 듣고 싶어 다시 요청드립니다
다음 요청시엔 머지하겠습니다 🙇

jinwoojwa committed:
refactor(product): 상품 생성 시 옵션을 포함하도록 변경

- Product.createProduct(...) 메서드를 추가해 상품과 옵션을 동시에 생성하도록 개선
- ProductService의 addProduct 로직에서 옵션 생성 책임 제거
- DataInitializer에서도 addOption() 대신 createProduct 사용

jinwoojwa commented:
한 주 동안 미션을 구현하는 데에만 급급했던 것 같습니다.
기능을 완성하는 데만 집중하느라 설계나 구조적인 부분에 대해 깊이 생각할 여유가 부족했는데 피드백을 받으면서 설계적인 부분부터 단일책임원칙, 그리고 N+1 문제 같은 중요한 개념들까지 다시 한번 깊게 고민해볼 수 있는 기회가 되었습니다.
기능 구현뿐 아니라 이러한 설계 원칙과 성능 이슈도 함께 고려할 수 있을 것 같습니다.
감사합니다!

liquidjoo reviewed (approved):
안녕하세요 진우님
피드백 반영 잘해주셨어요 👍 
같이 의견을 나누면서 즐거웠네요
저와 함께 했던 리뷰들이 도움이 되었길 빌며 마무리 하겠습니다
고생하셨습니다 🙇
